x/auth/legacy/v0_40/migrate.go

```diff
@@ -1,24 +1,133 @@
 package v040
 
 import (
-	"fmt"
+	pt "github.com/gogo/protobuf/types"
 
-	v038auth "github.com/cosmos/cosmos-sdk/x/auth/legacy/v0_38"
+	codectypes "github.com/cosmos/cosmos-sdk/codec/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
 	v039auth "github.com/cosmos/cosmos-sdk/x/auth/legacy/v0_39"
+	"github.com/cosmos/cosmos-sdk/x/auth/tx"
+	v040auth "github.com/cosmos/cosmos-sdk/x/auth/types"
+	v040vesting "github.com/cosmos/cosmos-sdk/x/auth/vesting/types"
 )
 
+// convertBaseAccount converts a 0.39 BaseAccount to a 0.40 BaseAccount.
+func convertBaseAccount(old *v039auth.BaseAccount) *v040auth.BaseAccount {
+	// In `x/auth/legacy/v0_38/migrate.go`, when creating a BaseAccount, we
+	// explicitly set the PublicKey field to nil. This propagates until 0.40,
+	// and we don't know the PubKey for those accounts, so we just put an empty
+	// string inside the Any.
+	var any *codectypes.Any
+	var err error
+	if old.PubKey != nil {
+		any, err = tx.PubKeyToAny(old.PubKey)
+	} else {
+		s := pt.StringValue{Value: ""}
+		any, err = codectypes.NewAnyWithValue(&s)
+	}
```

Comment:
Here's the output for BaseAccounts with nil pubkeys:

```js
     "auth": {
       "accounts": [
         {
           "@type": "/cosmos.auth.v1beta1.BaseAccount",
           "account_number": "0",
           "address": "cosmos1qqq9nqqllyvqfzen9w2sr0zp8jsfnycsjs8qwv",
           "pub_key": {
             "@type": "/google.protobuf.StringValue",
             "value": ""
           },
           "sequence": "1"
         },
// --snip--
```

is this okay? should I put an error message in the string value?

Reply:
Some context on the above comment.

In the current gaia hub, the genaccounts [don't have a `PubKey` field](https://github.com/cosmos/cosmos-sdk/blob/06b84d902a2038beef9927cc4eb1d72d5b1b35a9/x/genaccounts/legacy/v0_36/types.go#L14). When we convert that genesis to a more recent version (e.g. 0.37->0.38), the 0.38 migrate script can only create `BaseAccount` [with a nil `PubKey`](https://github.com/cosmos/cosmos-sdk/blob/6890feb3d28a1d98a8c8cd52669aca78361f5250/x/auth/legacy/v0_38/migrate.go#L16).

Continuing our genesis migration, we arrive at 0.40, where I show this protobuf String value in case there's no pubkey in the BaseAccount.

Reply:
what about using `google.protobuf.Empty`?

Reply:
Ah nice! I was looking around `Null` and `NullValue`, wasn't aware of empty!